lib/Sema/SemaExpr.cpp

```diff
@@ -12490,36 +12495,74 @@ ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
-    llvm::APInt I = llvm::APInt(1, 1, false);
-    uint64_t Bits = I.getZExtValue();
-    unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
-    llvm::APInt ResultVal(Width, Bits);
-    IntegerLiteral *One = IntegerLiteral::Create(
-        Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             One, SourceLocation());
-  } else if (ResE2.isUsable() && !ResE3.isUsable()) {
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             ResE2.get(), SourceLocation());
-  } else if (ResE2.isUsable() && ResE3.isUsable()) {
-    Result = ActOnRangeBoundsExpr(SourceLocation(), ResE2.get(), ResE3.get(),
-                                  SourceLocation());
+  switch (SyntaxType) {
+  case BoundsCastExpr::SyntaxType::Single:
+    if (DestTy->isCheckedPointerPtrType() || DestTy->isUncheckedPointerType()) {
+      llvm::APInt I = llvm::APInt(1, 1, false);
+      uint64_t Bits = I.getZExtValue();
+      unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
+      llvm::APInt ResultVal(Width, Bits);
+      IntegerLiteral *One = IntegerLiteral::Create(
+          Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
+      Result =
+          ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                               One, SourceLocation());
+    } else if (DestTy->isCheckedPointerArrayType()) {
+      Diag(TypeLoc,
+           diag::err_bounds_cast_array_pointer_type_with_single_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else if (DestTy->isIntegralType(Context)) {
+      Diag(TypeLoc, diag::err_bounds_cast_integral_with_single_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else
+      Result = CreateInvalidBoundsExpr();
+    break;
+  case BoundsCastExpr::SyntaxType::Count:
+    if (!DestTy->isCheckedPointerArrayType()) {
+      Diag(TypeLoc,
+           diag::err_bounds_cast_single_pointer_type_with_array_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else {
+      if (ResE2.isUsable())
+        Result = ActOnCountBoundsExpr(SourceLocation(),
+                                      BoundsExpr::Kind::ElementCount,
+                                      ResE2.get(), SourceLocation());
+    }
+    break;
+  case BoundsCastExpr::SyntaxType::Range:
+    if (!DestTy->isCheckedPointerArrayType()) {
+      Diag(TypeLoc,
+           diag::err_bounds_cast_single_pointer_type_with_array_syntax)
+          << DestTy;
+      Result = CreateInvalidBoundsExpr();
+    } else {
+      if (ResE2.isUsable() && ResE3.isUsable())
+        Result = ActOnRangeBoundsExpr(SourceLocation(), ResE2.get(),
+                                      ResE3.get(), SourceLocation());
+    }
+    break;
+  default:
```

Comment:
The default should have an llvm_unreachable call.   It is not expected for the code to reach this point.